### ISSUE TITLE
Slow FSGroup recursive permission changes cause customer confusion

### ISSUE THREAD

**Issue by mattcary** [kubernetes/kubernetes#126552]
### What happened?

[FSGroup volume permission setting](https://kubernetes.io/docs/tasks/configure-pod-container/security-context/#configure-volume-permission-and-ownership-change-policy-for-pods) is very useful for making NFS volumes just work in spite of the impedance between containers and traditional user id based linux file permissions. 

However, in practice with large volumes, fsgroup settings can cause workloads to effectively get stuck, as it can take hours or days to recursively change permissions on multi-terabyte servers. The fsGroupChangePolicy has helped that (onRootMismatch is a win), but this has still been a perennial source of customer issues.

The [kubelet does log when a long fsgroup change is in progress](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/volume_linux.go#L49), but this does not provide enough actionable information.

### What did you expect to happen?

The kubelet logging could be extended to provide updates of long-running fsgroup recursion, and maybe even some kind of progress (eg, some estimates of how much of the file tree has been walked).

### How can we reproduce it (as minimally and precisely as possible)?

Reproduction requires a large volume with many files, but it can be done with any volume type, eg RWO.

### Kubernetes version

<details>
All versions
</details>

### Cloud provider

<details>
We've seen customer complaints on GKE, but I imagine it happens on all providers.
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by mattcary** [comment]
> instead of just logging a warning after 30 seconds, we could implement periodic progress updates to give more visibility into how much of the file has been processed.

Yes, that's what I suggested in "What do you expect to happen"

**Comment by thisisharrsh** [comment]
```go
go func() {
    for range time.Tick(30 * time.Second) {
        progress := calculateProgress(processedFiles, totalFiles)
        klog.Infof("FSGroup progress: %.2f%% (%d/%d files)", progress, processedFiles, totalFiles)
        if progress >= 100 {
            return
        }
    }
}()
``` 
we can have an implementation like this.

**Comment by mattcary** [comment]
How do you get the total number of files? That by itself is expensive and could take O(minutes)? O(hours)? on a large disk.

**Comment by thisisharrsh** [comment]
Yes, you are right, I have just written a pseudo-function, we can discuss the final code and later have a PR for this issue.

**Comment by msau42** [comment]
BTW here was a previous attempt to improve the logging: https://github.com/kubernetes/kubernetes/pull/119359

**Comment by sunnylovestiramisu** [comment]
The previous fix attemped https://github.com/kubernetes/kubernetes/pull/119359 is to add a pod event for fsgroup taking a long time which is preferred than adding logs, it is more obvious to debug.

**Comment by thisisharrsh** [comment]
```go
// Increment file or directory count
if info.IsDir() {
	dirCount++
} else {
	fileCount++
}

// Periodically log progress every 1000 items or every 10 seconds
if (fileCount+dirCount)%1000 == 0 || time.Since(startTime).Seconds() > 10 {
	klog.V(2).Infof("FSGroup permission change in progress for %s: %d files, %d directories processed", dir, fileCount, dirCount)
	startTime = time.Now() // Reset timer for the next interval
}
```

**Comment by thisisharrsh** [comment]
In the implementation i have used `info.IsDir()` to determine if the current item is a directory or not. The `fileCount` and `dirCount` variables keep track of the total files and directories processed.

**Comment by thisisharrsh** [comment]
And then we are periodically logging is implemented to provide progress updates every 1000 items processed or every 10 seconds.

**Comment by thisisharrsh** [comment]
And then we are periodically logging to provide the progress updates every 1000 items processed or every 10 seconds.

**Comment by thisisharrsh** [comment]
Would like to hear from other members also if we have some better way to implement this.

**Comment by sunnylovestiramisu** [comment]
Are you also planning to send periodically events as well similar to the previous attempt in https://github.com/kubernetes/kubernetes/pull/119359?

**Comment by thisisharrsh** [comment]
Like my code, it executes the recursive traversal synchronously and logs progress based on processed file counts or time intervals. while the code in the [PR](https://github.com/kubernetes/kubernetes/pull/119359/files) runs the recursive operation in a separate goroutine and uses a ticker to log periodic warnings.

**Comment by thisisharrsh** [comment]
Another difference is that my code tracks and logs the number of files and directories processed at regular intervals while the code in the [PR](https://github.com/kubernetes/kubernetes/pull/119359/files) focuses on issuing warnings periodically if the operation takes too long.

**Comment by sunnylovestiramisu** [comment]
Any reason you prefer to do it synchronously instead of in a separate goroutine? A separate goroutine can have better performance and does not block the main application, but it may concurrent access files etc.

**Comment by sunnylovestiramisu** [comment]
Feedback from the sig-storage meeting, 1000 maybe too little, 10 seconds maybe too frequent. Because for filesystem the whole process can take 30mins etc. Maybe 1-3 mins is more appropriate.

If we have existing metrics to indicate how long this takes to verify.

**Comment by thisisharrsh** [comment]
> A separate goroutine can have better performance and does not block the main application, but it may concurrent access files etc.

I agree. 



> Any reason you prefer to do it synchronously instead of in a separate goroutine?

The primary reason for preferring synchronous execution is to avoids concurrency issues, as concurrent access to the same files or directories could lead to race conditions or inconsistencies.

**Comment by thisisharrsh** [comment]
> I think I like your idea of counting the files and updating progress from within the goroutine. I think that we do have to emit an event and not just a log -- not sure if that works from inside the goroutine, but you could always pass events out on a channel if it doesn't.

I agree that using a goroutine and emitting progress updates from within it can improve performance and responsiveness. 

I'll update the implementation to utilize a separate goroutine for file processing while sending progress updates and events through a channel.

**Comment by sunnylovestiramisu** [comment]
Is there a PR ready to be reviewed yet?

**Comment by thisisharrsh** [comment]
Raised the PR related to the issue [here.](https://github.com/kubernetes/kubernetes/pull/130227)
